lib/stages/ReadGain.cpp

```diff
@@ -0,0 +1,139 @@
+  //Apply config.
+  _num_elements = config.get<uint32_t>(unique_name, "num_elements");
+  scaling = config.get_default<float>(unique_name, "frb_scaling", 1.0);
+  vector<float> dg = {0.0, 0.0}; // re,im
+  default_gains = config.get_default<std::vector<float>>(unique_name, "frb_missing_gains", dg);
+
+  metadata_buf = get_buffer("network_buf");
+  register_consumer(metadata_buf, unique_name.c_str());
+  metadata_buffer_id = 0;
+  metadata_buffer_precondition_id = 0;
+  freq_idx = -1;
+  freq_MHz = -1;
+    
+  gain_buf = get_buffer("gain_buf");
+  gain_buf_id = 0;
+  register_producer(gain_buf, unique_name.c_str());
+  
+  update_gains = true;
+  first_pass = true;
+  
+  using namespace std::placeholders;
+
+  // listen for gain updates
+  _gain_dir = config.get_default<std::string>(unique_name, "updatable_config/gain_frb", "");
+  if (_gain_dir.length() > 0)
+    configUpdater::instance().subscribe(
+					config.get<std::string>(unique_name, "updatable_config/gain_frb"),
+					std::bind(&ReadGain::update_gains_callback, this, _1));
+}
+
+bool ReadGain::update_gains_callback(nlohmann::json& json) {
+    {
+      std::lock_guard<std::mutex> lock(mux);
+      update_gains = true;
+    }
+    cond_var.notify_all();
+    
+    try {
+        _gain_dir = json.at("frb_gain_dir");
+    } catch (std::exception& e) {
+        WARN("[FRB] Fail to read gain_dir {:s}", e.what());
+        return false;
+    }
+    INFO("[ReadGain] updated gain with {:s}============update_gains={:d}", _gain_dir, update_gains);
+    return true;
+
+}
+
+
+void ReadGain::main_thread() {
+
+  if (first_pass) {
+    first_pass = false;
+    uint8_t* frame = wait_for_full_frame(metadata_buf, unique_name.c_str(), metadata_buffer_precondition_id);
+    if (frame == NULL)
+      goto end_loop;
+    stream_id_t stream_id = get_stream_id_t(metadata_buf, metadata_buffer_id);
+    freq_idx = bin_number_chime(&stream_id);
+    freq_MHz = freq_from_bin(freq_idx);
+    //INFO("[CHECK ReadGain] freq_idx {:d} {:.2f}", freq_idx, freq_MHz);
+    metadata_buffer_precondition_id = (metadata_buffer_precondition_id + 1) % metadata_buf->num_frames;
+  }
+  mark_frame_empty(metadata_buf, unique_name.c_str(), metadata_buffer_id);
+  metadata_buffer_id = (metadata_buffer_id + 1) % metadata_buf->num_frames;
+  //unregister_consumer(metadata_buf, unique_name.c_str());
+    
+  while (!stop_thread) {
+    INFO("[ReadGain] update_gains={:d}-(0=false 1=true)-------------", update_gains);
+    {
+      std::unique_lock<std::mutex> lock(mux);
+      while (!update_gains) {
+	cond_var.wait(lock);//, [&]{return update_gains = true;});
+      }
+    }
+    INFO("[ReadGain] Going to update gain+++++++++++++++++update_gains={:d}", update_gains);
+
+    float* out_frame = (float*)wait_for_empty_frame(gain_buf, unique_name.c_str(), gain_buf_id);
+    if (out_frame == NULL) {
+      INFO("[ReadGain] waiting for frame={:d} but it is not empty", gain_buf_id);
```

Comment:
This isn't really a correct message.   If it returns `NULL` that's because the system is shutting down.